advanced_source/cpp_autograd.rst

```diff
@@ -0,0 +1,442 @@
+
+Create a tensor and set ``torch::requires_grad()`` to track computation with it
+
+.. code-block:: cpp
+
+  auto x = torch::ones({2, 2}, torch::requires_grad());
+  std::cout << x << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+  1 1
+  1 1
+  [ CPUFloatType{2,2} ]
+
+
+Do a tensor operation:
+
+.. code-block:: cpp
+
+  auto y = x + 2;
+  std::cout << y << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+   3  3
+   3  3
+  [ CPUFloatType{2,2} ]
+
+``y`` was created as a result of an operation, so it has a ``grad_fn``.
+
+.. code-block:: cpp
+
+  std::cout << y.grad_fn()->name() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+  AddBackward1
+
+Do more operations on ``y``
+
+.. code-block:: cpp
+
+  auto z = y * y * 3;
+  auto out = z.mean();
+  
+  std::cout << z << std::endl;
+  std::cout << z.grad_fn()->name() << std::endl;
+  std::cout << out << std::endl;
+  std::cout << out.grad_fn()->name() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+   27  27
+   27  27
+  [ CPUFloatType{2,2} ]
+  MulBackward1
+  27
+  [ CPUFloatType{} ]
+  MeanBackward0
+
+
+``.requires_grad_( ... )`` changes an existing Tensor's ``requires_grad`` flag in-place.
+
+.. code-block:: cpp
+
+  auto a = torch::randn({2, 2});
+  a = ((a * 3) / (a - 1));
+  std::cout << a.requires_grad() << std::endl;
+  
+  a.requires_grad_(true);
+  std::cout << a.requires_grad() << std::endl;
+  
+  auto b = (a * a).sum();
+  std::cout << b.grad_fn()->name() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+  false
+  true
+  SumBackward0
+
+Let's backprop now. Because ``out`` contains a single scalar, ``out.backward()``
+is equivalent to ``out.backward(torch::tensor(1.))``.
+
+.. code-block:: cpp
+
+  out.backward();
+
+Print gradients d(out)/dx
+
+.. code-block:: cpp
+
+  std::cout << x.grad() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+   4.5000  4.5000
+   4.5000  4.5000
+  [ CPUFloatType{2,2} ]
+
+You should have got a matrix of ``4.5``. For explanations on how we arrive at this value,
+please see `the corresponding section in this tutorial <https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html#gradients>`_.
+
+Now let's take a look at an example of vector-Jacobian product:
+
+.. code-block:: cpp
+
+  x = torch::randn(3, torch::requires_grad());
+  
+  y = x * 2;
+  while (y.norm().item<double>() < 1000) {
+    y = y * 2;
+  }
+    
+  std::cout << y << std::endl;
+  std::cout << y.grad_fn()->name() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+  -1021.4020
+    314.6695
+   -613.4944
+  [ CPUFloatType{3} ]
+  MulBackward1
+
+If we want the vector-Jacobian product, pass the vector to ``backward`` as argument:
+
+.. code-block:: cpp
+
+  auto v = torch::tensor({0.1, 1.0, 0.0001}, torch::kFloat);
+  y.backward(v);
+  
+  std::cout << x.grad() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+    102.4000
+   1024.0000
+      0.1024
+  [ CPUFloatType{3} ]
+
+You can also stop autograd from tracking history on Tensors with ``.requires_grad() == true``
```

Comment:
I don't think this reads as well as the python version.
I think you can simply state "on Tensors that require gradients".